modules/afmongodb/afmongodb.c

```diff
@@ -527,21 +528,45 @@ _free(LogPipe *d)
     mongoc_uri_destroy(self->uri_obj);
   if (self->coll_obj)
     mongoc_collection_destroy(self->coll_obj);
-  if (self->client)
-    mongoc_cleanup();
+
   log_threaded_dest_driver_free(d);
 }
 
 /*
  * Plugin glue.
  */
 
+static void
+afmongodb_global_init(void)
+{
+  mongoc_init();
+}
+
+static void
+afmongodb_global_deinit(void)
+{
+  mongoc_cleanup();
+}
+
+static void
+afmongodb_register_global_initializers(void)
+{
+  static gboolean initialized = FALSE;
+
+  if (!initialized)
+    {
+      register_application_hook(AH_RUNNING, (ApplicationHookFunc) afmongodb_global_init, NULL, AHM_RUN_ONCE);
```

Comment:
I think `AH_RUNNING` is too late to initialize libmongoc. Some libmongoc functions are already used in the `init` function, and worker threads are started earlier in `cfg_tree_on_inited()` at cfg init time.
According to libmongoc http://mongoc.org/libmongoc/1.16.1/init-cleanup.html
> It is responsible for initializing global state such as process counters, SSL, and threading primitives

I'm afraid we need a new state-like app-hook to solve this.

Reply:
AH_RUNNING worked in this case because apphook fires the freshly registered hook even if we register the hook after the specified AH_ state and miss the actual event.

Reply:
Sorry, I meant `AH_STARTUP`, not `AH_RUNNING`. I don't know why I ended up using `AH_RUNNING` instead, I'll check it again.

Reply:
Thank you very much for noticing this. `AH_STARTUP` should be used, I triple-checked.